### ISSUE TITLE
Better error handling when Travis fails

### ISSUE THREAD

**Issue by edgarcosta** [LMFDB/lmfdb#2414]
Improve the block
```
after_failure:
    # if some process failed, it will be clear in joblog which one failed
    - cat joblog
    # print all the stdout and stderr of nosetests run
    - find nosetests -type f -exec printf "\n\n filename = %s\n" {} \; -exec cat {} \;
```

perhaps first find the files with the string "FAIL:" or "... FAIL" and only display those, with some coloring.

**Comment by JohnCremona** [comment]
I am puzzled by this.  When I watch the test log as they happen, what I see is a few thousand lines scrolling past with success / failure for each one, until it gets to the end.  At that point if the exit code is 1 I can scroll back slowly (or start at the top) to look for a failure message.  Yet the script you quote above suggests that the job log is only displayed if there is a failure.  I see everything every time even when the tests pass.

**Comment by edgarcosta** [comment]
All that output comes from a single line
```
${SAGE} -sh -c 'parallel --joblog joblog --jobs 4 --results nosetests  --progress --verbose -a list ${SAGE_DIR}/local/bin/nosetests -v -s --testmatch=${TESTMATCH} ${COVERAGE}'
```

let me break this a bit.

Let's start by the end, for each line `X` in the file `list` it will run in parallel:
```
${SAGE_DIR}/local/bin/nosetests -v -s --testmatch=${TESTMATCH} ${COVERAGE} X
```

every line printed by the commands above is printed, but given that we are running everything in parallel these might get garbled. (the tool parallel tries to group the output and errors)

that issue leads us to `--results nosetests` this option saves the `stdout` and `stderr` for every `X`  in `nosetests/X/stdout` and `nosetests/X/stderr`.

for last the option` --joblog joblog`, creates a file named `joblog` where it reports on each command executed.
If one of the tests fails, we prints this file.
See for example: https://travis-ci.org/edgarcosta/lmfdb/builds/340317335
where on line 2560 reports that genus2_curve tests failed.

In the case a tests fails, we also print all the `nosetests/X/stdout` and `nosetests/X/stderr` in a "ungarbled" way with:
```
    # print all the stdout and stderr of nosetests run
    - find nosetests -type f -exec printf "\n\n filename = %s\n" {} \; -exec cat {} \;
```
See again: : https://travis-ci.org/edgarcosta/lmfdb/builds/340317335 (these start at line 2581)

Now back to the issue.
```
 When I watch the test log as they happen, what I see is a few thousand lines scrolling past with success / failure for each one, until it gets to the end.
```
what you are observing is due to the option `-v` in `nosetests`.
We could perhaps run `nosetests` with less verbosity, for example with `--quiet`, but that still produces a lot of output, see: https://travis-ci.org/edgarcosta/lmfdb/jobs/364297153

At the end of the way, I prefer more output and search for failed, than to have a minimal output that might not be very informative, for example, when the tests run out of time and and our session gets terminated, in that case, we will not run the `after_failure:` part and we will not know which test was holding everything back.


ps: the options `--progress --verbose` add a couple of lines that keep us posted regarding the progress of the tests as a whole, e.g:
```
local:4/0/100%/0.0s /home/travis/SageMath/local/bin/nosetests -v -s --testmatch="(?:^|\/)[Tt]est_" --with-coverage --cover-package=lmfdb lmfdb/raw.py
```

**Comment by edgarcosta** [comment]
Smaller update what I shoul do:
 If pyflakes failed make that clear.
 Only print the output of the tests that failed.

**Comment by JohnCremona** [comment]
I assume that this can be closed now that #2491 is merged?  I have not seen any buggy Travis output since though -- has anyone else?
